terraform/modules/cos/main.tf

```diff
@@ -76,6 +97,36 @@ resource "juju_integration" "mimir-grafana-source" {
   }
 }
 
+resource "juju_integration" "mimir-tracing-grafana-agent-traicing-provider" {
```

Comment:
For all juju_integration: Is there any situation where source/target application could be re-deployed?

If yes, then a lifecycle might be needed like the one in the documentation:
https://registry.terraform.io/providers/juju/juju/latest/docs/resources/integration

Reply:
Oh, the `lifecycle` key is interesting!

I think it depends on the scope of this `cos` module. For a first version I would think we don't need it.

WDYT @sed-i ?

Reply:
I think this comment from @amandahla is important. This `lifecycle` option in the `juju_integration` was added due to this issue:
- https://github.com/juju/terraform-provider-juju/issues/401

Which includes these steps to reproduce:
1. Deploy an application with relations via terraform.
2. Change its constraints and `terraform apply`
3. See the relations are lost, rerun `terraform apply` to restore relations.

Our modules support `constraints` and there may be other scenarios (none that I know of) where TF may remove an application and then the relation is missing unless someone manually runs `terraform apply`.

This would affect the robustness of our TF deployment if not implemented, although a rare scenario.

Reply:
We'll address this in a separate issue: https://github.com/canonical/observability/issues/235